source/scripts/model/DungeonGenerator.js

```diff
@@ -0,0 +1,504 @@
+     * Determines if another agent intersects this one.
+     * @param {Agent} agent - Agent test
+     * @return {bool} true if the other agent intersects this agent.
+     */
+    intersects(agent) {
+        let r1 = {
+            left: this.space.position.x,
+            top: this.space.position.y,
+            right: this.space.position.x + this.space.width,
+            bottom: this.space.position.y + this.space.height
+        }
+
+        let r2 = {
+            left: agent.space.position.x,
+            top: agent.space.position.y,
+            right: agent.space.position.x + agent.space.width,
+            bottom: agent.space.position.y + agent.space.height
+        }
+
+        return (r1.left < r2.right &&
+            r1.right > r2.left &&
+            r1.top > r2.bottom &&
+            r1.bottom < r2.top )
+
+        //return !(r2.left > r1.right ||
+        //   r2.right < r1.left ||
+        //   r2.top > r1.bottom ||
+        //   r2.bottom < r1.top);
+	}
+}
+
+/**
+ * Class to generate a random dungeon.
+ * @todo Fix trimming
+ * @todo Make more robust against islands
+ */
+class DungeonGenerator {
+    /**
+     * Constructs a random dungeon generator.
+     * @param {integer} maxSpaces - Maximum number of random spaces to generate.
+     * @param {width, height} minSpace - Minimum dimensions for a space.
+     * @param {width, height} maxSpace - Maximum dimensions for a space.
+     */
+    constructor(maxSpaces = null, minSpace = null, maxSpace = null) {
+        // arbitrary default value because I'm testing
+        this.maximumSpaces = maxSpaces || 30
+        this.minimumSpace = minSpace || {
+            width: 3,
+            height: 3
+        }
+        this.maximumSpace = maxSpace || {
+            width: 20,
+            height: 20
+        }
+        this.agents = new Array()
+
+        // flocking radius for the flocking behavior simulation
+        // can be changed to manipulate stuff
+        this.flockingRadius = 20
+
+        // flocking vector weights
+        // can be changed to manipulate flocking behavior
+        this.alignmentWeight = 2
+        this.cohesionWeight = 0.95
+        this.separationWeight = 1.1
+    }
+    /**
+     * Generates a random dungeon.
+     * @param {number} radius - Radius of random generation circle.
+     * @return {Array<Space>} Array of Space objects that represent the dungeon.
+     */
+    generate(radius) {
+        // would be cool to parallelize this
+        for (let i = 0; i < this.maximumSpaces; ++i) {
+            let dimensions = this.getRandomDimensions()
+
+            this.agents.push(new Agent(new Space({
+                position: this.getRandomPointInCircle(radius),
+                width: dimensions.width,
+                height: dimensions.height
+            })))
+        }
+
+        // I hate all of these side effects, but development time is short :(
+        this.flock()
+        this.cull()
+        this.buildGraph()
+        this.linkSpaces()
```

Comment:
It's fine; they're all within the same object, right? If the side-effects remain internal, they're really they're just mutators. So long as they're well documented, we're fine. :]